### PR TITLE
Merge API handlers

### LINKED ISSUE #641: Merge API handlers
Currently we have eight handlers related with retrieving resources from the database in the API:

```
	apisGroup.GET("/:group/:version/:resourceType", controller.GetAllResources)
	apisGroup.GET("/:group/:version/namespaces/:namespace/:resourceType", controller.GetNamespacedResources)
	apisGroup.GET("/:group/:version/namespaces/:namespace/:resourceType/:name", controller.GetNamespacedResourceByName)
	apisGroup.GET("/:group/:version/namespaces/:namespace/:resourceType/:name/log", controller.GetLogURLsByResourceName)

	apiGroup.GET("/:version/:resourceType", controller.GetAllCoreResources)
	apiGroup.GET("/:version/namespaces/:namespace/:resourceType", controller.GetNamespacedCoreResources)
	apiGroup.GET("/:version/namespaces/:namespace/:resourceType/:name", controller.GetNamespacedCoreResourceByName)
	apiGroup.GET("/:version/namespaces/:namespace/:resourceType/:name/log", controller.GetLogURLsByCoreResourceName)
```

The handlers of the URLS `/group/version/resourceType`, `/version/resourceType`, `/group/version/namespaces/namespace/resourceType` and `/version/namespaces/namespace/resourceType` are very similar. In order to simplify a little bit our code we could aggregate those handlers into one or two.

If we aggregate by database query, we can aggregate into two: resources and namespaced resources, as those use different queries `QueryResources` or `QueryNamespacedResources` resources respectively. This would be accomplished by merging the difference between core and non-core resources, which is the version/apiVersion creation. On the the core endpoints we pass the version as is to the database engine to be used as apiVersion on the DB. On the non-core endpoints we use group and version as apiVersion on the DB. A simple if clause could merge both:

```
        group := context.Param("group")
	version := context.Param("version")
	apiVersion := version
	if group != "" {
		apiVersion = fmt.Sprintf("%s/%s", group, version)
	}
        resources, lastId, lastDate, err = c.Database.QueryResources(context.Request.Context(), kind, apiVersion, limit, id, date)
```

If we aggregate also the namespaced endpoints we would need an if/else clause to use different queries, something like this:

```
        namespace := context.Param("namespace")
	if namespace != "" {
		resources, lastId, lastDate, err = c.Database.QueryNamespacedResources(context.Request.Context(), kind, apiVersion, namespace, limit, id, date)
	} else {
		resources, lastId, lastDate, err = c.Database.QueryResources(context.Request.Context(), kind, apiVersion, limit, id, date)
	}
```

**Acceptance Criteria**

* Merge handlers by database query, so merge core and non-core handlers (from 8 handlers to 4)
OR
* Merge handlers by return value: 1 list handler (core/non-core, namespaced/non-namespaced), 1 logs handler (core/non-core logs), 1 single resource handler (core/non-core by name).

### PR DESCRIPTION
**Pull request by rh-hemartin**
## Which Issue(s) this Pull Request Resolves

Resolves #641 

## Release Note

```release-note
NONE
```

## Notes for Reviewers

* Merged Database's `QueryNamespacedResources` and `QueryResources` so its more compact. I think the logic does not look bad in that case, let me know what do you think.

### CODE REVIEW

File: cmd/api/routers/routers.go [hunk 1]
```diff
@@ -41,43 +41,33 @@ type List struct {
...
+	}
+
+	var resources []*unstructured.Unstructured
+	var lastId int64
+	var lastDate string
+	if namespace != "" {
+		resources, lastId, lastDate, err = c.Database.QueryNamespacedResources(context.Request.Context(), kind, apiVersion, namespace, limit, id, date)
+	} else {
+		resources, lastId, lastDate, err = c.Database.QueryResources(context.Request.Context(), kind, apiVersion, limit, id, date)
 	}
```
Comment (maruiz93) [inline cmd/api/routers/routers.go:69]: I believe that we could make a similar merge of the Database `QueryNamespacedResources` and `QueryResources`, so we could replace this if/else block to a one call. 

I am ok doing that in a separate PR though.

**Review by maruiz93 — APPROVED** [review]

File: cmd/api/routers/routers.go (hunk 1 above)
Comment (rh-hemartin) [reply cmd/api/routers/routers.go]: I gave it a try, let me know if it looks good.

File: pkg/database/database.go [hunk 2]
```diff
@@ -50,8 +50,7 @@ type DBParamParser interface {
 }
 
 type DBInterface interface {
-	QueryResources(ctx context.Context, kind, apiVersion, limit, continueId, continueDate string) ([]*unstructured.Unstructured, int64, string, error)
-	QueryNamespacedResources(ctx context.Context, kind, apiVersion, namespace, limit, continueId, continueDate string) ([]*unstructured.Unstructured, int64, string, error)
+	QueryResources(ctx context.Context, kind, apiVersion, namespace, limit, continueId, continueDate string) ([]*unstructured.Unstructured, int64, string, error)
```
[the remark this answers is not in the document]
Comment (rh-hemartin) [reply pkg/database/database.go]: Good idea, I also implemented something that includes the name resources. Let me  know what you think.

File: pkg/database/database.go [hunk 3]
```diff
@@ -86,49 +84,32 @@ func (db *Database) Ping(ctx context.Context) error {
...
-
-	if continueId != "" && continueDate != "" {
-		pq = db.newParamQuery(db.info.GetNamespacedResourcesLimitedContinueSQL())
-		pq.addStringParams(kind, apiVersion, namespace, continueDate, continueId, limit)
-	} else {
+	if namespace != "" {
 		pq = db.newParamQuery(db.info.GetNamespacedResourcesLimitedSQL())
 		pq.addStringParams(kind, apiVersion, namespace, limit)
+
+		if name != "" {
```
Comment (ggallen) [inline pkg/database/database.go:96]: Can you only query using a `name` if you've used a `namespace` too?

Comment (rh-hemartin) [reply pkg/database/database.go:96]: Yes, to retrieve by name you need to specify namespace as we don't support cluster wide resources.

And yes, `limit` is not compatible with `name`. Retrieving by name always retrieve one element or it fails if there is more than one element.

File: cmd/api/routers/routers.go [hunk 4]
```diff
@@ -31,171 +31,64 @@ type Controller struct {
...
 
-	resource, err := c.Database.QueryNamespacedResourceByName(context.Request.Context(), kind, apiVersion, namespace, name)
-	if err != nil {
-		abort.Abort(context, err, http.StatusInternalServerError)
-		return
-	}
+	// We send namespace even if its an empty string (non-namespaced resources) the Database
+	// knows what to do
+	resources, lastId, lastDate, err := c.Database.QueryResources(context.Request.Context(), kind, apiVersion, namespace, name, limit, id, date)
```
Comment (maruiz93) [inline cmd/api/routers/routers.go:54]: We should be mindful with long lines,  line break suggestion and also added a couple of fixes in the comment
```suggestion
	// We send name and namespace even if it's an empty string (non-namespaced resources) the Database
	// knows what to do
	resources, lastId, lastDate, err := c.Database.QueryResources(context.Request.Context(), 
	    kind, apiVersion, namespace, name, limit, id, date)
```

File: pkg/database/database.go [hunk 5]
```diff
@@ -50,9 +50,7 @@ type DBParamParser interface {
 }
 
 type DBInterface interface {
-	QueryResources(ctx context.Context, kind, apiVersion, limit, continueId, continueDate string) ([]string, int64, string, error)
-	QueryNamespacedResources(ctx context.Context, kind, apiVersion, namespace, limit, continueId, continueDate string) ([]string, int64, string, error)
-	QueryNamespacedResourceByName(ctx context.Context, kind, apiVersion, namespace, name string) (string, error)
+	QueryResources(ctx context.Context, kind, apiVersion, namespace, name, limit, continueId, continueDate string) ([]string, int64, string, error)
```
Comment (maruiz93) [inline pkg/database/database.go:53]: Be mindful to the line length:
```suggestion
	QueryResources(ctx context.Context, 
	    kind, apiVersion, namespace, name, limit, continueId, continueDate string) ([]string, int64, string, error)
```

File: pkg/database/fake/database_generated.go [hunk 6]
```diff
@@ -102,8 +84,17 @@ func (f *Database) QueryLogURLs(ctx context.Context, kind, apiVersion, namespace
 	return urls, f.err
 }
 
-func (f *Database) QueryNamespacedResources(ctx context.Context, kind, version, namespace, limit, continueId, continueDate string) ([]string, int64, string, error) {
-	resources := f.filterResourcesByKindApiVersionAndNamespace(kind, version, namespace)
+func (f *Database) QueryResources(ctx context.Context, kind, version, namespace, name, limit, continueId, continueDate string) ([]string, int64, string, error) {
```
Comment (maruiz93) [inline pkg/database/fake/database_generated.go:87]: line break:
```suggestion
func (f *Database) QueryResources(ctx context.Context, 
    kind, version, namespace, name, limit, continueId, continueDate string) ([]string, int64, string, error) {
```

**Review by maruiz93 — APPROVED** [review]

**Review by ggallen — APPROVED** [review]

File: cmd/api/routers/routers.go (hunk 4 above)
Comment (rh-hemartin) [reply cmd/api/routers/routers.go]: Done

**Pull request merged**